packages/langium/src/index/ast-descriptions.ts

```diff
@@ -0,0 +1,113 @@
+/******************************************************************************
+ * Copyright 2021 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { LangiumDocument } from '../documents/document';
+import { Linker } from '../references/linker';
+import { NameProvider } from '../references/naming';
+import { AstNodeDescription } from '../references/scope';
+import { LangiumServices } from '../services';
+import { AstNode } from '../syntax-tree';
+import { AstNodeReference, getDocument, streamAllContents, streamContents, streamReferences } from '../utils/ast-util';
+import { AstNodeLocator } from './ast-node-locator';
+
+export interface ReferenceDescription {
+    sourceUri: string // URI of the document that holds a reference
+    sourcePath: string // Path to AstNode that holds a reference
+    targetUri: string // target document uri
+    targetPath: string // how to find target AstNode inside the document
+    start: number
+    end: number
+}
+
+export interface AstNodeDescriptionProvider {
+    createDescription(node: AstNode, name: string, document: LangiumDocument): AstNodeDescription;
+    createDescriptions(document: LangiumDocument): AstNodeDescription[];
+}
+
+export interface ReferenceDescriptionProvider {
+    createDescriptions(document: LangiumDocument): ReferenceDescription[];
+}
+
+export class DefaultAstNodeDescriptionsProvider implements AstNodeDescriptionProvider {
+
+    protected readonly astNodeLocator: AstNodeLocator;
+    protected readonly nameProvider: NameProvider;
+
+    constructor(services: LangiumServices) {
+        this.astNodeLocator = services.index.AstNodeLocator;
+        this.nameProvider = services.references.NameProvider;
+    }
+
+    createDescription(node: AstNode, name: string, document: LangiumDocument): AstNodeDescription {
+        return {
+            node,
+            name,
+            type: node.$type,
+            documentUri: document.textDocument.uri,
+            path: this.astNodeLocator.getAstNodePath(node)
+        };
+    }
+
+    createDescriptions(document: LangiumDocument): AstNodeDescription[] {
+        const descr: AstNodeDescription[] = [];
+        const rooNode = document.parseResult?.value;
+        if (rooNode) {
+            const name = this.nameProvider.getName(rooNode);
+            if (name) {
+                descr.push(this.createDescription(rooNode, name, document));
+            }
+            streamContents(rooNode).forEach(content => {
+                const name = this.nameProvider.getName(content.node);
+                if (name) {
+                    descr.push(this.createDescription(content.node, name, document));
```

Comment:
The index should not contain references to the actual AST nodes. Maybe we need two different variants of `createDescription`, one for local scope and one for global scope?

Reply:
I'm not sure we should introduce a new interface for that, but it is definitely something we should discuss in a new issue. 
Currently we set `AstNodeDescription.node` to undefined before adding the AstNodeDescription into the index, which is fine so far.